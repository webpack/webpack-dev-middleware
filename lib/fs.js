'use strict';

const fs = require('fs');
const path = require('path');
const MemoryFileSystem = require('memory-fs');
const pathabs = require('path-is-absolute');
const { chalk } = require('webpack-log');
const NodeOutputFileSystem = require('webpack/lib/node/NodeOutputFileSystem');
const DevMiddlewareError = require('./DevMiddlewareError');

const { mkdirp } = new NodeOutputFileSystem();

module.exports = {
  toDisk(context) {
    context.compiler.hooks.afterEmit.tap('WebpackDevMiddleware', (compilation) => {
      const { assets, compiler } = compilation;
      const { log } = context;
      const { writeToDisk: filter } = context.options;
      let { outputPath } = compiler;

      if (outputPath === '/') {
        outputPath = compiler.context;
      }

      for (const assetPath of Object.keys(assets)) {
        const asset = assets[assetPath];
        const source = asset.source();
        const isAbsolute = pathabs(assetPath);
        const writePath = isAbsolute ? assetPath : path.join(outputPath, assetPath);
        const relativePath = path.relative(process.cwd(), writePath);
        const allowWrite = filter && typeof filter === 'function' ? filter(writePath) : true;

        if (allowWrite) {
          let output = source;

          mkdirp.sync(path.dirname(writePath));

          if (Array.isArray(source)) {
            output = source.join('\n');
          }

          try {
            fs.writeFileSync(writePath, output, 'utf-8');
            log.debug(chalk`{cyan Asset written to disk}: ${relativePath}`);
          } catch (e) {
            log.error(`Unable to write asset to disk:\n${e}`);
          }
        }
      }
    });
  },

  setFs(context, compiler) {
    if (typeof compiler.outputPath === 'string' && !pathabs.posix(compiler.outputPath) && !pathabs.win32(compiler.outputPath)) {
      throw new DevMiddlewareError('`output.path` needs to be an absolute path or `/`.');
    }

    let fileSystem;
    // store our files in memory
    const isMemoryFs = !compiler.compilers && compiler.outputFileSystem instanceof MemoryFileSystem;

    if (isMemoryFs) {
      fileSystem = compiler.outputFileSystem;
    } else {
      fileSystem = new MemoryFileSystem();
      compiler.outputFileSystem = fileSystem;
    }

    context.fs = fileSystem;
  }
};
