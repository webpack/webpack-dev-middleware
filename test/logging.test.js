import fs from 'fs';
import path from 'path';

import execa from 'execa';
import webpack from 'webpack';

import clearDirectory from './helpers/clearDirectory';

function extractWebpackEntry(string) {
  const matches = string.match(/webpack\s\d\.\d\d?.\d\d?/gim);

  const result =
    matches === null
      ? null
      : matches.map((i) =>
          i
            .replace(/\d\d\./g, 'xx.')
            .replace(/\d\./g, 'xx.')
            .replace(/\.\d\d/g, '.xx')
            .replace(/\.\d/g, '.xx')
        )[0];

  return result;
}

function extractCountCompilations(string) {
  const matches = webpack.webpack
    ? string.match(/webpack\s\d\.\d\d?.\d\d?/gim)
    : string.match(/Entrypoint/gim);

  const result = matches === null ? null : matches.length;

  return `Count compilations: ${result}`;
}

function extractErrorEntry(string) {
  const matches = string.match(/code:\s['||"]?\w.+['||"]/gim);

  return matches;
}

describe('logging', () => {
  it('should logging on successfully build', (done) => {
    const runner = `${__dirname}/helpers/runner.js`;

    let proc;

    try {
      proc = execa(runner, [], {
        stdio: 'pipe',
        env: {
          WC: 'webpack.config',
          WCF_stats: 'normal',
        },
      });
    } catch (error) {
      throw error;
    }

    let stdOut = '';
    let stdError = '';

    proc.stdout.on('data', (chunk) => {
      stdOut += chunk.toString();

      if (!/error/gi.test(stdOut)) {
        proc.stdin.write('|exit|');
      }
    });

    proc.stderr.on('data', (chunk) => {
      stdError += chunk.toString();
      proc.stdin.write('|exit|');
    });

    proc.on('exit', () => {
      expect(
        `${extractWebpackEntry(stdOut)}\n${extractCountCompilations(stdOut)}`
      ).toMatchSnapshot('data');
      expect(stdError).toMatchSnapshot('error');

      done();
    });
  });

  it('should logging on successfully build in multi-compiler mode', (done) => {
    const runner = `${__dirname}/helpers/runner.js`;

    let proc;

    try {
      proc = execa(runner, [], {
        stdio: 'pipe',
        env: {
          WC: 'webpack.array.config',
        },
      });
    } catch (error) {
      throw error;
    }

    let stdOut = '';
    let stdError = '';

    proc.stdout.on('data', (chunk) => {
      stdOut += chunk.toString();

      if (!/error/gi.test(stdOut)) {
        proc.stdin.write('|exit|');
      }
    });

    proc.stderr.on('data', (chunk) => {
      stdError += chunk.toString();
      proc.stdin.write('|exit|');
    });

    proc.on('exit', () => {
      expect(
        `${extractWebpackEntry(stdOut)}\n${extractCountCompilations(stdOut)}`
      ).toMatchSnapshot('data');
      expect(stdError).toMatchSnapshot('error');

      done();
    });
  });

  it('should logging on unsuccessful build', (done) => {
    const runner = `${__dirname}/helpers/runner.js`;

    let proc;

    try {
      proc = execa(runner, [], {
        stdio: 'pipe',
        env: {
          WC: 'webpack.error.config',
        },
      });
    } catch (error) {
      throw error;
    }

    let stdOut = '';
    let stdError = '';

    proc.stdout.on('data', (chunk) => {
      stdOut += chunk.toString();

      if (/error/gi.test(stdOut)) {
        proc.stdin.write('|exit|');
      }
    });

    proc.stderr.on('data', (chunk) => {
      stdError += chunk.toString();
      proc.stdin.write('|exit|');
    });

    proc.on('exit', () => {
      expect(
        `${extractWebpackEntry(stdOut)}\n${extractCountCompilations(stdOut)}`
      ).toMatchSnapshot('data');
      expect(stdError).toMatchSnapshot('error');

      done();
    });
  });

  it('should logging on unsuccessful build in multi-compiler', (done) => {
    const runner = `${__dirname}/helpers/runner.js`;

    let proc;

    try {
      proc = execa(runner, [], {
        stdio: 'pipe',
        env: {
          WC: 'webpack.array.error.config',
        },
      });
    } catch (error) {
      throw error;
    }

    let stdOut = '';
    let stdError = '';

    proc.stdout.on('data', (chunk) => {
      stdOut += chunk.toString();

      if (/error/gi.test(stdOut)) {
        proc.stdin.write('|exit|');
      }
    });

    proc.stderr.on('data', (chunk) => {
      stdError += chunk.toString();
      proc.stdin.write('|exit|');
    });

    proc.on('exit', () => {
      expect(
        `${extractWebpackEntry(stdOut)}\n${extractCountCompilations(stdOut)}`
      ).toMatchSnapshot('data');
      expect(stdError).toMatchSnapshot('error');

      done();
    });
  });

  it('should logging an warning', (done) => {
    const runner = `${__dirname}/helpers/runner.js`;

    let proc;

    try {
      proc = execa(runner, [], {
        stdio: 'pipe',
        env: {
          WC: 'webpack.warning.config',
        },
      });
    } catch (error) {
      throw error;
    }

    let stdOut = '';
    let stdError = '';

    proc.stdout.on('data', (chunk) => {
      stdOut += chunk.toString();

      if (/warning/gi.test(stdOut)) {
        proc.stdin.write('|exit|');
      }
    });

    proc.stderr.on('data', (chunk) => {
      stdError += chunk.toString();
      proc.stdin.write('|exit|');
    });

    proc.on('exit', () => {
      expect(
        `${extractWebpackEntry(stdOut)}\n${extractCountCompilations(stdOut)}`
      ).toMatchSnapshot('data');
      expect(stdError).toMatchSnapshot('error');

      done();
    });
  });

  it('should logging warnings in multi-compiler mode', (done) => {
    const runner = `${__dirname}/helpers/runner.js`;

    let proc;

    try {
      proc = execa(runner, [], {
        stdio: 'pipe',
        env: {
          WC: 'webpack.array.warning.config',
        },
      });
    } catch (error) {
      throw error;
    }

    let stdOut = '';
    let stdError = '';

    proc.stdout.on('data', (chunk) => {
      stdOut += chunk.toString();

      if (/warning/gi.test(stdOut)) {
        proc.stdin.write('|exit|');
      }
    });

    proc.stderr.on('data', (chunk) => {
      stdError += chunk.toString();
      proc.stdin.write('|exit|');
    });

    proc.on('exit', () => {
      expect(
        `${extractWebpackEntry(stdOut)}\n${extractCountCompilations(stdOut)}`
      ).toMatchSnapshot('data');
      expect(stdError).toMatchSnapshot('error');

      done();
    });
  });

  it('should logging an error in "watch" method', (done) => {
    const runner = `${__dirname}/helpers/runner.js`;

    let proc;

    try {
      proc = execa(runner, [], {
        stdio: 'pipe',
        env: {
          WCF_infrastructureLogging_level: 'log',
          WATCH_break: true,
        },
      });
    } catch (error) {
      throw error;
    }

    let stdError = '';

    proc.stderr.on('data', (chunk) => {
      stdError += chunk.toString();
      proc.stdin.write('|exit|');
    });

    proc.on('exit', () => {
      expect(extractErrorEntry(stdError)).toMatchSnapshot('error');

      done();
    });
  });

  it('should logging an error from the fs error when the "writeToDisk" option is "true"', async (done) => {
    const runner = `${__dirname}/helpers/runner.js`;
    const outputDir = path.resolve(
      __dirname,
      './outputs/write-to-disk-mkdir-error'
    );

    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir);
    }

    fs.chmodSync(outputDir, 0o444);

    let proc;

    try {
      proc = execa(runner, [], {
        stdio: 'pipe',
        env: {
          WC: 'webpack.simple.config',
          WCF_output_filename: 'bundle.js',
          WCF_output_path: outputDir,
          WCF_infrastructureLogging_level: 'log',
          WMC_writeToDisk: true,
          FS_block: true,
        },
      });
    } catch (error) {
      throw error;
    }

    let stdError = '';

    proc.stderr.on('data', (chunk) => {
      stdError += chunk.toString();
      proc.stdin.write('|exit|');
    });

    proc.on('exit', () => {
      expect(extractErrorEntry(stdError)).toMatchSnapshot('error');

      fs.chmodSync(outputDir, 0o777);
      clearDirectory(outputDir);

      done();
    });
  });
});
