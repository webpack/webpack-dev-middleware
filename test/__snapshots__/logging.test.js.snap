// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logging should logging an error from the fs error when the "writeToDisk" option is "true": error 1`] = `
Array [
  "code: 'EACCES'",
]
`;

exports[`logging should logging an error in "watch" method: error 1`] = `
Array [
  "code: 'watch error'",
]
`;

exports[`logging should logging an warning: data 1`] = `
"webpack xx.xx.xx
Count compilations: 1"
`;

exports[`logging should logging an warning: error 1`] = `""`;

exports[`logging should logging on successfully build in multi-compiler mode: data 1`] = `
"webpack xx.xx.xx
Count compilations: 2"
`;

exports[`logging should logging on successfully build in multi-compiler mode: error 1`] = `""`;

exports[`logging should logging on successfully build: data 1`] = `
"webpack xx.xx.xx
Count compilations: 1"
`;

exports[`logging should logging on successfully build: error 1`] = `""`;

exports[`logging should logging on unsuccessful build in multi-compiler: data 1`] = `
"webpack xx.xx.xx
Count compilations: 2"
`;

exports[`logging should logging on unsuccessful build in multi-compiler: error 1`] = `""`;

exports[`logging should logging on unsuccessful build: data 1`] = `
"webpack xx.xx.xx
Count compilations: 1"
`;

exports[`logging should logging on unsuccessful build: error 1`] = `""`;

exports[`logging should logging warnings in multi-compiler mode: data 1`] = `
"webpack xx.xx.xx
Count compilations: 2"
`;

exports[`logging should logging warnings in multi-compiler mode: error 1`] = `""`;
